nixos/modules/services/continuous-integration/buildkite-agents.nix

```diff
@@ -184,20 +185,34 @@ in
       };
     };
   };
-
-  config = mkIf config.services.buildkite-agent.enable {
-    users.users.buildkite-agent = {
-      name = "buildkite-agent";
+  enabledAgents = lib.filterAttrs (n: v: v.enable) cfg;
+  mapAgents = function: lib.mkMerge (lib.mapAttrsToList function enabledAgents);
+in
+{
+  options.services.buildkite-agents = mkOption {
+    type = types.attrsOf (types.submodule buildkiteOptions);
+    default = {};
+    description = ''
+      Attribute set of buildkite agents.
+      The attribute key is combined with the hostname and a unique integer to
+      create the final agent name. This can be overridden by setting the `name`
+      attribute.
+    '';
+  };
+  config.users.users = mapAgents (name: cfg: {
+    "buildkite-agent-${name}" = {
+      name = "buildkite-agent-${name}";
       home = cfg.dataDir;
       createHome = true;
       description = "Buildkite agent user";
       extraGroups = [ "keys" ];
       isSystemUser = true;
     };
-
-    environment.systemPackages = [ cfg.package ];
-
-    systemd.services.buildkite-agent =
+  });
+  # TODO: specifying multiple different buildkite packages would conflict
```

Comment:
This conflict should probably be handled explicitly, maybe with an assertion, and also documented. Also also, why have separate options if you absolutely must specify the same value for each?

Reply:
Separate options seem useful for running different versions/patches of the buildkite agent, but I'm not sure what to do with the `systemPackages` in that case.

Reply:
I don't really see many reasons for `buildkite-agent` to be available in `systemPackages` - I see it useful to be in `runtimePackages`, so build jobs can append steps to the currently running job, but that should be it…